src/main/java/org/infinispan/persistence/mongodb/MongoDbStore.java

```diff
@@ -0,0 +1,193 @@
+            databaseName = connString.getDatabase();
+        } else {
+            databaseName = connectionConfiguration.database();
+        }
+        String collectionName = connectionConfiguration.collection();
+        MongoDatabase database = mongoClient.getDatabase(databaseName);
+        this.collection = database.getCollection(collectionName);
+
+        return CompletableFuture.completedFuture(null);
+    }
+
+    @Override
+    public CompletionStage<Void> stop() {
+        mongoClient.close();
+        return CompletableFuture.completedFuture(null);
+    }
+
+
+    @Override
+    public CompletionStage<Long> size(IntSet segments) {
+        return Single.fromPublisher(collection.countDocuments()).toCompletionStage();
+    }
+
+    @Override
+    public CompletionStage<Long> approximateSize(IntSet segments) {
+        return size(segments);
+    }
+
+
+    @Override
+    public CompletionStage<Boolean> containsKey(int segment, Object key) {
+        return load(segment, key).thenApply(Objects::nonNull);
+    }
+
+    @Override
+    public CompletionStage<MarshallableEntry<K, V>> load(int segment, Object key) {
+        long now = context.getTimeService().wallClockTime();
+        Object storeKey = cacheToStoreConverter.toStoreKey(key);
+        Bson filter = and(eq("_id", storeKey), or(gt(EXPIRY_TIME, now), lte(EXPIRY_TIME, -1L)));
+
+        return Maybe.fromPublisher(collection.find(filter))
+                .map(document -> cacheToStoreConverter.toCacheEntry(key, document))
+                .toCompletionStage(null);
+    }
+
+    @Override
+    public CompletionStage<Void> write(int segment, MarshallableEntry<? extends K, ? extends V> marshallableEntry) {
+        Object storeKey = cacheToStoreConverter.toStoreKey(marshallableEntry.getKey());
+        Document storeEntry = cacheToStoreConverter.toStoreEntry(storeKey, marshallableEntry);
+
+        Publisher<UpdateResult> publisher = collection.replaceOne(eq("_id", storeKey), storeEntry, new ReplaceOptions().upsert(true));
+        return Single.fromPublisher(publisher)
+                .ignoreElement()
+                .toCompletionStage(null);
+    }
+
+    @Override
+    public CompletionStage<Boolean> delete(int segment, Object key) {
+        BasicDBObject query = new BasicDBObject();
+        query.put("_id", cacheToStoreConverter.toStoreKey(key));
+
+        return Single.fromPublisher(collection.deleteOne(query))
+                .map(e -> e.getDeletedCount() > 0)
+                .toCompletionStage();
+    }
+
+    @Override
+    public CompletionStage<Void> clear() {
+        return Completable.fromPublisher(collection.drop())
+                .toCompletionStage(null);
+    }
+
+
+    @Override
+    public Publisher<MarshallableEntry<K, V>> publishEntries(IntSet segments, Predicate<? super K> filter, boolean includeValues) {
+        return Flowable.defer(() -> {
+            FindPublisher<Document> iterable = collection.find();
```

Comment:
We may want to use the `StoreConfiguration#maxBatchSize` here for the `FindPublisher.batchSize`. I don't know how large of a batch the server would normally send back.

Reply:
Added.